### ISSUE TITLE
loops Scope is exponentially slower in 0.2.8 than 0.1.55

### ISSUE THREAD

**Issue by zachary-jablons-okcupid** [google/jax#5496]
Hey there,

I've been using JAX (on CPU) for a while now, and I was recently running a benchmark I hadn't run in a long time and found that I was getting much worse performance than I remember. I setup an environment with 0.1.55 (the JAX version I used when I last ran this benchmark), and found that there's been a significant slowdown in this benchmark when comparing the latest version (0.2.8) with the older version!

![image](https://user-images.githubusercontent.com/72406994/105558899-ca825900-5cdd-11eb-87fc-09aecd18f03d.png)

This code/benchmark relies on encapsulating a training epoch in `loops.Scope` and then compiling it to get the speed observed. While I guess I'm not 100% confident that the issue is `loops.Scope` slowing down, it seems like the likely culprit. The relevant snippet is here:

```
@partial(jit, static_argnums=(7, 8))                          
def train_epoch(vem, tem, vbv, tbv,                           
                voter_indices, target_indices, ratings,       
                batch_size, batched_dataset_size):            
                                        
    with loops.Scope() as s:                                  
        s.vem = vem                                           
        s.tem = tem                                           
        s.vbv = vbv                                           
        s.tbv = tbv                                           
        for i in s.range(0, batched_dataset_size):            
            batch_start = i * batch_size                      
            voter_ib = lax.dynamic_slice(                     
                voter_indices, (batch_start,), (batch_size,)) 
            target_ib = lax.dynamic_slice(                    
                target_indices, (batch_start,), (batch_size,))
            ratings_b = lax.dynamic_slice(                    
                ratings, (batch_start,), (batch_size,))       
            s.vem, s.tem, s.vbv, s.tbv = train_batch(         
                s.vem, s.tem, s.vbv, s.tbv,                   
                voter_ib, target_ib, ratings_b)               
        return s.vem, s.tem, s.vbv, s.tbv                     
return train_epoch                                            
```

Relevant packages in the new environment:
```
jax==0.2.8
jaxlib==0.1.59
```

Relevant packages in the old environment:
```
jax==0.1.55
jaxlib==0.1.37
```

Both are in Python 3.7 (not sure if that matters). 

*I don't have a fully reproducible self contained example yet (working on it)* but I wanted to file this issue in the meantime in case it's related to a known issue or the result of a deliberate design decision (or I should change something I'm doing).

Notes:

- I've verified that the results I get from each version are the same.
- This may be related to [this issue](https://github.com/google/jax/issues/2491)?
- I haven't bisected to figure out when this regression may have occurred, so it may have happened before 0.2.8. It's a bit of a pain to figure out the correct jax/jaxlib/numpy versions for each jax version.

**Comment by mattjj** [comment]
Thanks for the clear explanation!

I wonder if this is related to [omnistaging](https://github.com/google/jax/blob/master/design_notes/omnistaging.md), which can lead to more code getting staged out to XLA, which then in turn can hit XLA codegen bugs that lead to exponential compile times.

To check if this is the issue, can you try [disabling omnistaging](https://github.com/google/jax/blob/master/design_notes/omnistaging.md#how-can-i-disable-omnistaging-for-now) and check if the regression is still present? It's as easy as calling `jax.config.disable_omnistaging()` at the top of your file, or setting the `JAX_OMNISTAGING=0` shell environment variable.

**Comment by zachary-jablons-okcupid** [comment]
Thanks for the quick response! 

Just tried that, and unfortunately disabling omnistaging does not appear to make any sort of difference. Don't have a self contained reproduction yet but here's some more detailed statistics:

```
Omnistaging enabled: False
Using jax==0.2.8
nvotes range:   0%|                                                                                    | 0/6 [00:00<?, ?it/s]
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
nvotes range: 100%|████████████████████████████████████████████████████████████████████████████| 6/6 [00:24<00:00,  4.13s/it]
   n_samples      total    extract     inner  test_norm
0        100   1.635107   0.000317  1.633622   1.502444
1        500   1.062631   0.001376  1.059045   3.052246
2       1000   0.989900   0.004927  0.982001   4.454413
3       5000   1.159750   0.157465  0.989604   9.878271
4      10000   1.694304   0.675206  0.995073  14.074360
5      50000  18.169523  16.876715  1.037657  31.681110
```

Note that `total` is the total time taken (all of these are in seconds), `extract` is the time taken to turn the `DeviceArray` output into an ndarray (using `numpy.array`), and `inner` is the time taken to 'run' the jit functions (including, I think, the compilation time itself?). So, this is for 0.2.8, but for 0.1.55 I'm seeing:

```
Using jax==0.1.55
nvotes range:   0%|                                                                                    | 0/6 [00:00<?, ?it/s]
/home/u1/zach/.pyenv/versions/jax_old/lib/python3.7/site-packages/jax/lib/xla_bridge.py:120: UserWarning: No GPU/TPU found, f
alling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
nvotes range: 100%|████████████████████████████████████████████████████████████████████████████| 6/6 [00:08<00:00,  1.45s/it]
   n_samples     total   extract     inner  test_norm
0        100  1.848387  0.000430  1.846686   1.502444
1        500  1.467144  0.000559  1.464809   3.052246
2       1000  1.345396  0.001296  1.340998   4.454413
3       5000  1.272209  0.007389  1.252182   9.878271
4      10000  1.280177  0.006150  1.249469  14.074359
5      50000  1.429890  0.030495  1.288265  31.681108
```

Notably these two perform pretty similarly up to ~50000 samples where there's a dramatic increase in time (although the newer version does take significantly longer before that).

**Comment by hawkinsp** [comment]
I think we'd need a reproduction to make progress here.

Also note that the `loops` support is experimental!

**Comment by zachary-jablons-okcupid** [comment]
Alright, so I've ventured a bit into the depths of reproducing this and lost a bit of sanity on the way. There's _something_ weird going on here in the old version of JAX where it actually seems to take a bit of work to get this speedup but it appears to depend on something I didn't expect - the loss function?

Here's the [gist](https://gist.github.com/zachary-jablons-okcupid/cc8af3161552208b2a0e1c1611983b76), which is a self-contained example but not minimal. I tried to build a minimal example from first principles here (training a univariate model, then multivariate, then single embedding, then dual embedding ([something like this](https://sifter.org/~simon/journal/20061211.html)) which is very close to the actual code I'm seeing the issue with) but was unable to reproduce the issue until I replaced the loss function with the one we actually use. The nature of this was that _both_ versions of jax were fairly sluggish, but when I moved the initial loss from:

```
def linear(w, v, y,):       
    return y - forward(w, v)
```

to

```
def l2_loss(w, v, y):          
    return 0.5 * jnp.power(    
            y - forward(w,v), 2
    )                          
```

The old version of jax sped right up (from `7.760422229766846 seconds to train 1000 samples for 5 epochs` to `0.027247190475463867 seconds to train 1000 samples for 5 epochs`) whereas the new version did not (it actually slowed down, from `3.466555118560791 seconds to train 1000 samples for 5 epochs` to `12.952296733856201 seconds to train 1000 samples for 5 epochs`).

In the attached gist there's four loss functions, and the only one that does not display this behavior is the `linear` one - the others all speed up in the old version. I'm not quite sure what the difference is that's causing this, as the `linear` method is still wrapped in a `jnp.mean` at some point?

I'm still planning on getting to a more minimal reproduction (without the dual embedding) but I figured I'd post this now to see if there's something obvious to the jax team that I'm not seeing.

I've verified (again) here that the results from each version are the same. Let me know if there's anything else I can help with to figure out this issue!

**Comment by zachary-jablons-okcupid** [comment]
Hey there,

Just bumping this issue as I've come back to it in my work. Tested with `jax==0.2.10` and `jaxlib==0.1.62`, still seeing a similar slowdown.

Thanks in advance for any help you can offer

**Comment by hawkinsp** [comment]
We've dropped `jax.experimental.loops`, so I think this issue is moot.

**Comment by zachary-jablons-okcupid** [comment]
Should I replace the experimental loops with e.g. [`fori_loop`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.fori_loop.html) and expect similar performance?

**Comment by hawkinsp** [comment]
Yes, I'd expect similar behavior.
